Initial version of Tuple implementation.

TysonAndre opened the pull request:
**I still need to add proper exception handling.** The implementation is also inefficient (currently treating integer positions as strings)

This involved more code than I expected.

(e.g. still need to add checks for ranges of positional arguments in Tuple)

Code coverage is incomplete, not fully sure of what methods TupleValue should expose

@lucaswoj what was UnspecifiedTupleType needed for? Are these the right things to start implementing?

TysonAndre committed:
Initial version of Tuple implementation.

Still buggy.
(e.g. still need to add checks for ranges of positional arguments)

Code coverage is incomplete, not fully sure of how TupleValue is going to be
used.

lucaswoj commented:
Thanks for getting started on this Tyson. This is gonna be a pretty big project. Perhaps we can break this up into more manageable tasks -- why don't you finish `TupleTypeValue` before starting on `TupleValue`

I'm ok with the inefficiency of treating positional arguments as strings for now. We don't need to worry about performance yet and this keeps the code conceptually simple for the time being.

An unspecified tuple is a type declaration that says "I am a tuple but my exact type is not known at compile time" -- `(Integer baz, Boolean bar)  foo` vs `Tuple foo`. This will be needed for some moremetaprogrammey features but can be deferred for a later pull request.

Always mention the issue number somewhere in the commit message or PR. This creates a nice link between the ticket and pull request. fixes #24

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Attribute.java:
```diff
@@ -0,0 +1,63 @@
+package org.whdl.frontend.syntaxtree;
+
+
+public class Attribute {
```
If you're note comfortable nesting this class under the `TupleValue` class then please rename it to `TupleValueAttribute`. Just calling it `Attribute` is a little vague for my taste.

lucaswoj replied:
That being said, feel free to leave all the `TupleValue` classes out of this commit for the time being.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Attribute.java:
```diff
@@ -0,0 +1,63 @@
+    TypeValue actualType = value.getType();
+    if(!actualType.isSubtypeOf(expectedType)) {
+      throw new TypeMismatchException(expectedType, actualType);
+    }
```
This code should be abstracted out into `TypeValue`. If it isn't already, please move it there.

maxqchen replied:
I'm pretty sure expected type here should just be type, not type.getType() which always returns TypeTypeValue.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  From: lucaswojSent: Thursday, January 30, 2014 7:54 AMTo: whdl/whdlReply To: whdl/whdlSubject: Re: [whdl] Initial version of Tuple implementation. (#59)In src/main/java/org/whdl/frontend/syntaxtree/Attribute.java:

> -  
> -  public AttributeType getType() {
> -    return this.type;
> -  }
> -  
> -  public void verify() throws Exception {
> -    if(value == null) {
> -      throw new AttributeNotDefinedException(type);
> -    }
> -    value.verify();
>   +
> -    TypeValue expectedType = type.getType();
> -    TypeValue actualType = value.getType();
> -    if(!actualType.isSubtypeOf(expectedType)) {
> -      throw new TypeMismatchException(expectedType, actualType);
> -    }

This code should be abstracted out into TypeValue. If it isn't already, please move it there. 

—Reply to this email directly or view it on GitHub.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/AttributeAccessException.java:
```diff
@@ -0,0 +1,15 @@
+package org.whdl.frontend.syntaxtree;
+
+public class AttributeAccessException extends Exception {
```
I think this name could be more explicit. What's the exception? It looks like the exception is thrown when an attribute is accessed but that attribute does not exist. Perhaps `TupleAttributeNotDefined` would be best for consistency with `VariableNotDefinedException`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/AttributeDefinitionException.java:
```diff
@@ -0,0 +1,19 @@
+package org.whdl.frontend.syntaxtree;
+
+public class AttributeDefinitionException extends Exception {
```
`TupleAttributeMultipleDefinitionException`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/AttributeNotDefinedException.java:
```diff
@@ -0,0 +1,14 @@
+package org.whdl.frontend.syntaxtree;
+
+public class AttributeNotDefinedException extends Exception {
```
This sounds like it should be `TupleAttributeNotAssignedException`

Its interesting how there are so many commonalities between this and Scope. Perhaps there's a way to factor out this functionality... I can't think of a logical way to do it, though.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/AttributeType.java:
```diff
@@ -0,0 +1,96 @@
+package org.whdl.frontend.syntaxtree;
+
+
+public class AttributeType {
```
Either nest this class inside `TupleType` or rename it `TupleTypeAttributeType`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/AttributeType.java:
```diff
@@ -0,0 +1,96 @@
+  private final String name;
+  private final boolean positional;
+  
+  // final String destructuringName; // not sure if this is part of a tuple
```
Yes, I think this should be here for the time being.

TysonAndre replied:
What is a destructuringName?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/BitValue.java:
```diff
@@ -13,7 +13,7 @@ public static BitValue getInstance(boolean value) {
   }
 
-  private boolean value;
+  private final boolean value;
```
Sweet. Thanks! In fact, because everything is single assignment, we should probably do this more liberally. I'll open a separate issue for this

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+
+public class SpecifiedTupleTypeValue extends TupleTypeValue implements Iterable<AttributeType> {
+	private Map<String, AttributeType> attributeTypes = new HashMap<String, AttributeType>();
+	private AttributeType duplicatedAttribute = null;
```
What is this?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+	    // TODO(tyson) check for duplicates in verification
+	    AttributeType old = attributeTypes.put(a.getName(), a);
+	    if(old != null) {
+	      duplicatedAttribute = old;
```
This seems awkward... Why not throw the exception right away? Why not require it gets passed in as a `Map` to avoid the possibility of duplicates?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+	  for(AttributeType a : attrTypes) {
+	    // TODO(tyson) check for duplicates in verification
+	    AttributeType old = attributeTypes.put(a.getName(), a);
+	    if(old != null) {
```
There should be a space after keywords -- `if (` vs `if(` (This seems to be the case for the next 10ish lines)

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+	}
+	
+	public boolean isSubtypeOf(TypeValue other) {
+	  if (other instanceof SpecifiedTupleTypeValue) {
```
This is actually going to be a fair bit more complicated in the long run... This is ok for now though :+1:

lucaswoj replied:
Actually, we might want to have a couple methods, one asking if the two tuple types are defined EXACTLY the same and another asking if one type can be cast to the other.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+
+  @Override
+  public void verify() throws Exception {
+    if(duplicatedAttribute != null) {
```
Watch out for the space after keywords `if (`. I'll add a linter soon...

lucaswoj replied:
I don't like this system of keeping `duplicatedAttribute` as a member variable. See the constructor for potential solutions

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+    if(duplicatedAttribute != null) {
+      throw new AttributeDefinitionException(this, duplicatedAttribute.getName());
+    } else if (attributeTypes.isEmpty()) {
+      throw new AttributeDefinitionException(this, null);
```
I don't think a null tuple should be an exception.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/SpecifiedTupleTypeValue.java:
```diff
@@ -0,0 +1,82 @@
+      if(!attributeTypes.containsKey(Integer.toString(i))) {
+        throw new AttributeDefinitionException(this, i);
+      }
+    }
```
What's the point of this check? You constructed the positional arguments, you should be reasonably sure there are as many as you set! The `verify` method is for catching user errors and doing type checks, not general assertions.

TysonAndre replied:
This is checking if they're continuous. That's probably unlikely, since the user can't specify the position of positional arguments explicitly

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleTypeValue.java:
```diff
@@ -0,0 +1,5 @@
+package org.whdl.frontend.syntaxtree;
+
+public abstract class TupleTypeValue extends TypeValue {
```
For now, feel free to rename `SpecifiedTupleTypeValue` to `TupleTypeValue`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleValue.java:
```diff
@@ -0,0 +1,127 @@
+ * how it would be handled.
+ */
+public class TupleValue extends Value implements Iterable<Attribute> {
+  // TODO: make an array of Attributes for positional parameters?
```
No. Better to keep things simple in the short term and worry about performance later.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleValue.java:
```diff
@@ -0,0 +1,127 @@
+public class TupleValue extends Value implements Iterable<Attribute> {
+  // TODO: make an array of Attributes for positional parameters?
+	private Map<String, Attribute> attributes = new HashMap<String, Attribute>();
+	private SpecifiedTupleTypeValue typeValue; // may be reassigned
```
When might this be reassigned?

TysonAndre replied:
I forget why I wrote this, it was probably early on.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleValue.java:
```diff
@@ -0,0 +1,127 @@
+          e.printStackTrace();
+        }
+		  } else {
+		    v = a.getDefaultValue();
```
What if `a` doesn't have a default value?

TysonAndre replied:
that would probably be a compile-time error, and I'd have to throw an exception in the constructor instead...
getDefaultValue() would return null

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleValue.java:
```diff
@@ -0,0 +1,127 @@
+ * TODO(tyson): figure out what UnspecifiedTupleTypeValue would be needed for,
+ * how it would be handled.
+ */
+public class TupleValue extends Value implements Iterable<Attribute> {
```
Maybe consider a separate pull request for `TupleValue` and related classes

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleValue.java:
```diff
@@ -0,0 +1,127 @@
+	 * Construct a TupleValue from a list of attributes
+	 * @param attrs a list of attributes
+	 */
+	public TupleValue(Collection<Attribute> attrs) {
```
This should take a `TupleTypeValue` instead of a list of attributes. For one thing, we need to assign `this.typeValue`

TysonAndre replied:
As far as I understand, you can only assign a value once, so a tuple would need all of its fields when it was constructed.

That constructor signature would create an uninitialized tuple

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TupleValue.java:
```diff
@@ -0,0 +1,127 @@
+
+	/**
+	 * A tuple can be synthesized if each attributes is synthesizable.
+	 * TODO: how does this work with Value requiring isSynthesizable() || isCompiletimeEvaluable()
```
Interesting question. We should discuss this soon.

mtrberzi commented:
Good work so far. This is a difficult one but you've made great progress. I'll go over the code in more detail once you've had a chance to address Lucas's fixes and review some of the notes you left for yourself (e.g. adding proper exception handling, which I think is the only thing stopping the build at the moment).

TysonAndre committed:
More work on lucas's comments

removed work on Tuple values.

Added more coverage to tests, toString functions.

Also removed hashCode() from TupleAttributeType since it wouldn't be used.
(Not going to make a Set<Attributes> and look up by a new Attribute())

TysonAndre committed:
Allow verify to throw exception in Expression

I assume verify communicates by throwing Exceptions...

TysonAndre committed:
remove unused exception type.

TysonAndre commented:
@MurphEngineer - merged with master and fixed that a few days ago - Expression.verify() should throw Exception (or a subtype).

Also, I'm going on vacation, will be back in a week on the 24th of February.

lucaswoj commented:
This is probably a bad time to bring this up, but I'm starting to feel uncomfortable with the hybrid dict/array datatype. More advanced ways of accessing tuple members just gets so gross...
